frame/treasury/src/lib.rs

```diff
@@ -66,33 +66,70 @@ pub mod weights;
 use codec::{Decode, Encode, MaxEncodedLen};
 use scale_info::TypeInfo;
 
-use sp_runtime::{
-	traits::{AccountIdConversion, CheckedAdd, Saturating, StaticLookup, Zero},
-	Permill, RuntimeDebug,
-};
-use sp_std::{collections::btree_map::BTreeMap, prelude::*};
-
+use codec::FullCodec;
 use frame_support::{
-	print,
+	dispatch::fmt::Debug,
 	traits::{
-		Currency, ExistenceRequirement::KeepAlive, Get, Imbalance, OnUnbalanced,
-		ReservableCurrency, WithdrawReasons,
+		tokens::{AssetId, Balance, BalanceConversion},
+		Get,
 	},
 	weights::Weight,
 	PalletId,
 };
+use sp_runtime::{
+	traits::{CheckedAdd, Saturating, StaticLookup, Zero},
+	Permill, RuntimeDebug,
+};
+use sp_std::{collections::btree_map::BTreeMap, prelude::*};
 
 pub use pallet::*;
 pub use weights::WeightInfo;
 
-pub type BalanceOf<T, I = ()> =
-	<<T as Config<I>>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
-pub type PositiveImbalanceOf<T, I = ()> = <<T as Config<I>>::Currency as Currency<
-	<T as frame_system::Config>::AccountId,
->>::PositiveImbalance;
-pub type NegativeImbalanceOf<T, I = ()> = <<T as Config<I>>::Currency as Currency<
-	<T as frame_system::Config>::AccountId,
->>::NegativeImbalance;
+/// Status for making a payment via the `Pay::pay` trait function.
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Encode, Decode, Eq, PartialEq, Clone, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+pub enum PaymentStatus {
+	/// Payment is in progress. Nothing to report yet.
+	InProgress,
+	/// Payment status is unknowable. It will never be reported successful or failed.
+	Unknown,
+	/// Payment happened successfully.
+	Success,
+	/// Payment failed. It may safely be retried.
+	Failure,
+}
+
+// TODO: Should be replaced with a version in frame_support
+pub trait Pay {
+	/// The type by which we measure units of the currency in which we make payments.
+	type Balance: Balance;
+	/// The type by which we identify the individuals to whom a payment may be made.
+	type AccountId;
+	/// An identifier given to an individual payment.
+	type Id: FullCodec + MaxEncodedLen + TypeInfo + Clone + Eq + PartialEq + Debug + Copy;
+	type AssetId: AssetId;
+	/// Make a payment and return an identifier for later evaluation of success in some off-chain
+	/// mechanism (likely an event, but possibly not on this chain).
+	fn pay(
+		who: &Self::AccountId,
+		asset_id: Self::AssetId,
```

Comment:
I'd call it something different to `AssetId` since that is well-defined in XCM and we might want it to mean something a bit different. E.g. we might want to use a type of `(MultiLocation, AssetId)` to represent the chain on which the asset resides and then the `AssetId` for the asset on that chain.

```suggestion
		asset_kind: Self::AssetKind,
```